src/main/java/org/folio/service/transactions/batch/BatchTransactionService.java

```diff
@@ -0,0 +1,229 @@
+    if (transactionPatches.isEmpty()) {
+      return;
+    }
+    throw new HttpException(500, "transactionPatches: not implemented");
+  }
+
+  private Future<Void> applyChanges(BatchTransactionHolder holder, DBConn conn) {
+    return createTransactions(holder, conn)
+      .compose(v -> updateTransactions(holder, conn))
+      .compose(v -> deleteTransactions(holder, conn))
+      .compose(v -> updateBudgets(holder, conn));
+  }
+
+  private Future<Void> createTransactions(BatchTransactionHolder holder, DBConn conn) {
+    List<Transaction> transactions = holder.getAllTransactionsToCreate();
+    if (transactions.isEmpty()) {
+      return succeededFuture();
+    }
+    return transactionDAO.createTransactions(transactions, conn)
+      .onSuccess(ids -> logger.info("Batch transactions: successfully created {} transactions", transactions.size()))
+      .onFailure(t -> logger.error("Batch transactions: failed to create transactions, transactions = {}",
+        transactions, t))
+      .mapEmpty();
+  }
+
+  private Future<Void> updateTransactions(BatchTransactionHolder holder, DBConn conn) {
+    List<Transaction> transactions = holder.getAllTransactionsToUpdate();
+    if (transactions.isEmpty()) {
+      return succeededFuture();
+    }
+    return transactionDAO.updateTransactions(transactions, conn)
+      .onSuccess(v -> logger.info("Batch transactions: successfully updated {} transactions", transactions.size()))
+      .onFailure(t -> logger.error("Batch transactions: failed to update transactions, transactions = {}",
+        transactions, t));
+  }
+
+  private Future<Void> deleteTransactions(BatchTransactionHolder holder, DBConn conn) {
+    List<String> ids = holder.getIdsOfTransactionsToDelete();
+    if (ids.isEmpty()) {
+      return succeededFuture();
+    }
+    return transactionDAO.deleteTransactionsByIds(ids, conn)
+      .onSuccess(v -> logger.info("Batch transactions: successfully deleted {} transactions", ids.size()))
+      .onFailure(t -> logger.error("Batch transactions: failed to delete transactions, ids = {}", ids, t));
+  }
+
+  private Future<Void> updateBudgets(BatchTransactionHolder holder, DBConn conn) {
+    List<Budget> budgets = holder.getBudgets();
+    if (budgets.isEmpty()) {
+      return succeededFuture();
+    }
+    return budgetService.updateBatchBudgets(budgets, conn)
+      .onSuccess(v -> logger.info("Batch transactions: successfully updated {} budgets", budgets.size()))
+      .onFailure(t -> logger.error("Batch transactions: failed to update budgets, budgets = {}", budgets, t))
+      .mapEmpty();
+  }
+
+  private void populateMetadata(Batch batch, Map<String, String> okapiHeaders) {
+    // NOTE: Okapi usually populates metadata when a POST method is used with an entity with metadata.
+    // But in the case of the batch API there is no top-level metadata, so it is not populated automatically.
+    String userId = okapiHeaders.get(XOkapiHeaders.USER_ID);
+    if (userId == null) {
```

Comment:
I prefer for such cases always use StringUtils.isNotBlank() to not deep dive into Okapi implementation can userId be populated as empty String

Reply:
I copied that from the Okapi implementation, I would prefer to keep it the same (we might need to update it at some point).